gammapy/datasets/flux_points.py

```diff
@@ -122,11 +139,52 @@ def __init__(
         self.models = models
         self.meta_table = meta_table
 
-        if mask_safe is None:
-            mask_safe = (~data.is_ul).data
+        self.available_stat_type = dict(
+            chi2=self._stat_array_chi2,
+            profile=self._stat_array_profile,
+            distrib=self._stat_array_distrib,
+        )
+
+        if stat_kwargs is None:
+            stat_kwargs = dict()
+        self.stat_kwargs = stat_kwargs
 
+        self.stat_type = stat_type
+
+        if mask_safe is None:
+            mask_safe = np.ones(self.data.dnde.data.shape, dtype=bool)
         self.mask_safe = mask_safe
 
+    @property
+    def stat_type(self):
+        return self._stat_type
+
+    @stat_type.setter
+    def stat_type(self, stat_type):
+
+        if stat_type not in self.available_stat_type:
+            raise ValueError(
+                "Invalid stat_type: possible options are {list(self.available_stat_type.keys())}"
+            )
+
+        if stat_type == "chi2":
+            self.mask_valid = (~self.data.is_ul).data & np.isfinite(self.data.dnde)
```

Comment:
Both `mask_safe` and `mask_valid` are necessary?

Reply:
Please make `mask_valid` a lazy property then, to make it read only.

Reply:
done